realm/app-manager/src/launcher/oci.rs

```diff
@@ -0,0 +1,351 @@
+    RaTlsCertResolverCreation(#[source] RaTlsError),
+
+    #[error("Failed to create OCI client")]
+    OciClientCreation(#[source] ir_client::error::Error),
+
+    #[error("Failed to fetch image info from image registry")]
+    ImageInfoFetching(#[source] ir_client::error::Error),
+
+    #[error("Version is invalid: {0}")]
+    InvalidVersion(#[source] ir_client::error::Error, String),
+
+    #[error("Failed to download and unpack image")]
+    UnpackError(#[source] ir_client::error::Error),
+
+    #[error("Required image annotations weren't found")]
+    FailedToFindRequiredAnnotations(),
+
+    #[error("Metadata not found")]
+    MetadataNotFound(),
+
+    #[error("Runtime config is missing in image config")]
+    RuntimeConfigMissing(),
+
+    #[error("Entrypoint and cmd are missing")]
+    EntryPointCmdMissing(),
+
+    #[error("Group or User name {0} is missing")]
+    InvalidUserOrGroup(String),
+
+    #[error("Failed to resolve user name or group name")]
+    GetPwName(#[source] Errno),
+
+    #[error("Path convertsion error, {0:?} is not a valid Path")]
+    PathConversionError(#[source] Infallible, String),
+}
+
+pub struct OciLauncher {
+    launcher_config: OciLauncherConfig,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Metadata {
+    vendor_cert: Vec<Vec<u8>>,
+    config_hash: String,
+    image_config: OciConfig,
+}
+
+impl OciLauncher {
+    pub fn from_oci_config(launcher_config: OciLauncherConfig) -> Self {
+        Self { launcher_config }
+    }
+
+    fn create_oci_client(&self, im_url: &str) -> Result<Client> {
+        let config = Config::builder().host(im_url.to_owned());
+
+        let oci_client = match &self.launcher_config {
+            OciLauncherConfig::NoTLS => Client::from_config(config.no_tls()),
+
+            OciLauncherConfig::RusTLS { root_ca } => Client::from_config(config.rustls_no_auth(
+                load_root_cert_store(root_ca).map_err(OciLauncherError::RootCaLoading)?,
+            )),
+
+            OciLauncherConfig::RaTLS {
+                root_ca,
+                token_resolver,
+            } => {
+                let token_resolver = match token_resolver {
+                    TokenResolver::FromFile(path) => TokenFromFile::from_path(path)
+                        .map_err(OciLauncherError::TokenReadingError)?,
+                };
+
+                let cert_resolver =
+                    RaTlsCertResolver::from_token_resolver(Arc::new(token_resolver))
+                        .map_err(OciLauncherError::RaTlsCertResolverCreation)?;
+
+                Client::from_config(config.ratls(
+                    load_root_cert_store(root_ca).map_err(OciLauncherError::RootCaLoading)?,
+                    Arc::new(cert_resolver),
+                ))
+            }
+        };
+
+        Ok(oci_client.map_err(OciLauncherError::OciClientCreation)?)
+    }
+
+    async fn try_read_metadata(&self, path: &Path) -> Option<Metadata> {
+        read_to_string(path.join("metadata.json"))
```

Comment:
Maybe create an impl level CONST with this `"metadata.json"` str? It is repeating in two different methods.